wallet/src/Pos/Wallet/Web/Methods/Payment.hs

```diff
@@ -160,38 +159,44 @@ sendMoney SendActions{..} passphrase moneySource dstDistr = do
     addrMetas <- nonEmpty addrMetas' `whenNothing`
         throwM (RequestError "Given money source has no addresses!")
 
-    sks <- forM addrMetas $ getSKByAddress passphrase
     srcAddrs <- forM addrMetas $ decodeCTypeOrFail . cwamId
-
-    withSafeSigners sks (pure passphrase) $ \ss -> do
-        let hdwSigner = NE.zip ss srcAddrs
-
-        relatedAccount <- getSomeMoneySourceAccount moneySource
-        outputs <- coinDistrToOutputs dstDistr
-        (th, dstAddrs) <-
-            rewrapTxError "Cannot send transaction" $ do
-                (txAux, inpTxOuts') <-
-                    prepareMTx hdwSigner outputs (relatedAccount, passphrase)
-
-                ts <- Just <$> getCurrentTimestamp
-                let tx = taTx txAux
-                    txHash = hash tx
-                    inpTxOuts = toList inpTxOuts'
-                    dstAddrs  = map txOutAddress . toList $
-                                _txOutputs tx
-                    th = THEntry txHash tx Nothing inpTxOuts dstAddrs ts
-                ptx <- mkPendingTx srcWallet txHash txAux th
-
-                (th, dstAddrs) <$ submitAndSaveNewPtx enqueueMsg ptx
-
-        logInfoS $
-            sformat ("Successfully spent money from "%
-                     listF ", " addressF % " addresses on " %
-                     listF ", " addressF)
-            (toList srcAddrs)
-            dstAddrs
-
-        addHistoryTx srcWallet th
+    let metasAndAdrresses = zip (toList addrMetas) (toList srcAddrs)
+
+    let getSinger addr = do
+          let addrMeta =
+                  fromMaybe (error "Corresponding adress meta not found")
+                            (fst <$> find ((== addr) . snd) metasAndAdrresses)
+          sk <- getSKByAddress passphrase addrMeta
```

Comment:
Ok, great, so this is part of the crucial change. We now only call `getSKByAddress` for selected addresses.